tracer.go

```diff
@@ -239,27 +240,20 @@ func (r *tracerImpl) RecordSpan(raw RawSpan) {
 
 // Flush sends all buffered data to the collector.
 func (r *tracerImpl) Flush() {
+	r.flushingLock.Lock()
+	defer r.flushingLock.Unlock()
 	r.lock.Lock()
+	defer r.lock.Unlock()
```

Comment:
This is dangerous. `defer r.lock.Unlock()` implies to other programmers that the entire function is inside the `r.lock`, which is not the case, and causes an extra call to `Unlock()` in one case, which is very dangerous. (you unlock once, other code locks, you unlock again, race occurs). 

I would suggest extracting two functions, `beforeFlush() (*reportBuffer, bool) ` and `afterFlush(error)`, which encapsulate the two sets of buffer/memory management. That way, in each function you can use `defer r.lock.Unlock()` and it's cleaner.